java/common/exception/ErrorMessage.java

```diff
@@ -53,20 +53,28 @@
             new ErrorMessage(16, "Invalid define/undefine query. User defined variables are not accepted in define/undefine query.");
     public static final ErrorMessage INVALID_RULE_WHEN_MISSING_PATTERNS =
             new ErrorMessage(17, "Rule '%s' 'when' has not been provided with any patterns.");
+    public static final ErrorMessage INVALID_RULE_WHEN_NESTED_NEGATION =
+            new ErrorMessage(18, "Rule '%s' 'when' contains a nested negation.");
+    public static final ErrorMessage INVALID_RULE_WHEN_CONTAINS_DISJUNCTION =
+            new ErrorMessage(19, "Rule '%s' 'when' contains a disjunction.");
+    public static final ErrorMessage INVALID_RULE_THEN_ZERO_CONSTRAINTS =
+            new ErrorMessage(20, "Rule '%s' 'then' must contain at least one constraint: '%s'.");
     public static final ErrorMessage INVALID_RULE_THEN_ONE_CONSTRAINT =
-            new ErrorMessage(18, "Rule '%s' 'then' with one constraint must be a new (or extended) relation, attribute ownership, or 'isa' downcast: '%s'.");
+            new ErrorMessage(21, "Rule '%s' 'then' with one constraint must be an attribute ownership: '%s'.");
     public static final ErrorMessage INVALID_RULE_THEN_TWO_CONSTRAINTS =
-            new ErrorMessage(19, "Rule '%s' 'then' with two constraints must must be a relation and an isa constraint: '%s'.");
+            new ErrorMessage(22, "Rule '%s' 'then' with two constraints must must be a relation and an isa constraint: '%s'.");
+    public static final ErrorMessage INVALID_RULE_THEN_THREE_OR_MORE_CONSTRAINT =
+            new ErrorMessage(23, "Rule '%s' 'then' must cannot contain more than three constraint: '%s'.");
```

Comment:
Should we really split this error message to 3 different error message? How do you know that the user intended to put only 1 or 2 specifically? Telling them that they can only insert attribute ownership for 1 constraint, and relation an `isa` for 2 constraints, assume that those are the user's intentions. I think we should keep 1 generic error message for `then` violations.

Reply:
Good point. I was thinking in terms of the code and not the way the user will see it.